creator-node/src/services/URSMRegistrationManager.js

```diff
@@ -0,0 +1,215 @@
+const axios = require('axios')
+const _ = require('lodash')
+const { Utils: LibsUtils } = require('@audius/libs')
+
+const { logger } = require('../logging')
+const { generateTimestampAndSignature } = require('../apiSigning')
+const { parseCNodeResponse } = require('../apiHelpers')
+
+const NumSignaturesRequired = 3
+
+/**
+ * This Service is responsible for registering this node on the UserReplicaSetManager L2 contract (URSM)
+ *
+ * @notice Service is backwards compatible, and will work before and after URSM contract deployment
+ */
+class URSMRegistrationManager {
+  constructor (nodeConfig, audiusLibs) {
+    this.nodeConfig = nodeConfig
+    this.audiusLibs = audiusLibs
+
+    this.delegateOwnerWallet = nodeConfig.get('delegateOwnerWallet')
+    this.delegatePrivateKey = nodeConfig.get('delegatePrivateKey')
+    this.spOwnerWallet = nodeConfig.get('spOwnerWallet')
+    this.isUserMetadataNode = nodeConfig.get('isUserMetadataNode')
+
+    if (!this.isUserMetadataNode &&
+      (!this.audiusLibs || !this.delegateOwnerWallet || !this.delegatePrivateKey || !this.spOwnerWallet)
+    ) {
+      throw new Error('URSMRegistrationManager cannot start due to missing required configs')
+    }
+  }
+
+  logInfo (msg) {
+    logger.info(`URSMRegistrationManager || ${msg}`)
+  }
+
+  logError (msg) {
+    logger.error(`URSMRegistrationManager ERROR || ${msg}`)
+  }
+
+  /**
+   * Registers node on UserReplicaSetManager contract (URSM)
+   *
+   * Steps:
+   *  1. Fetch node record from L1 ServiceProviderFactory for spID
+   *    a. Short-circuit if no L1 record found
+   *  2. Fetch node record from L2 UserReplicaSetManager for spID
+   *    a. Short-circuit if L2 record for node already matches L1 record (i.e. delegateOwnerWallets match)
+   *  3. Fetch list of all nodes registered on URSM, in order to submit requests for proposal
+   *    a. Randomize list to minimize bias + de-dupe by ownerWallet to meet on-chain uniqueness constraint
+   *  4. Request signatures from all registered nodes, in batches of 3, until 3 successful signatures received
+   *    a. Error if all available nodes contacted without 3 successful signatures
+   *  5. Submit registration transaction to URSM with signatures
+   */
+  async run () {
+    this.logInfo('Beginning URSM registration process')
+
+    if (this.isUserMetadataNode) {
+      this.logInfo('URSMRegistration cannot run in userMetadataNode')
+      return
+    }
+
+    /**
+     * (Backwards-compatibility) Short circuit if L2 URSM contract not yet deployed
+     *
+     * TODO - currently this will retry until contract is deployed, definitely not ideal
+     */
+    if (!this.audiusLibs.contracts.UserReplicaSetManagerClient) {
```

Comment:
i feel like the `init` could be divorced from the `run` aspect of this flow, since we later will have to rely on a similar condition for node processing but not a blocker imo